README
/***************************************************************************/
#	 	spcaview: Sdl video recorder and viewer with sound.         #
#		spcaserv Streaming TCP server				    #
#		spcacat Picture grabber					    #
#This package work with the spca5xx based webcam with the raw jpeg feature. #
#All the decoding is in user space with the embedded jpeg decoder/encoder   #
#.                                                                          #
# 		Copyright (C) 2003 2004 2005 Michel Xhaard                  #
#                                                                           #
# This program is free software; you can redistribute it and/or modify      #
# it under the terms of the GNU General Public License as published by      #
# the Free Software Foundation; either version 2 of the License, or         #
# (at your option) any later version.                                       #
#                                                                           #
# This program is distributed in the hope that it will be useful,           #
# but WITHOUT ANY WARRANTY; without even the implied warranty of            #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             #
# GNU General Public License for more details.                              #
#                                                                           #
# You should have received a copy of the GNU General Public License         #
# along with this program; if not, write to the Free Software               #
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA #
#                                                                           #
/***************************************************************************/

CREDITS:
--------
Original code
	from sview the testing tool for the spca50x v4l drivers. Made and copyrighted
	by Till Adam, Miah Gregory, and Francois Beerten in 2003.
	All rights reserved.
SDL_audioin A smart library to get sound input writted by:
 *   Tyler Montbriand 
 *   tsm@accesscomm.ca 
 *   The library are linked internally in spcaview
Avilib  Some utilities for writing and reading AVI files.
 *  These are not intended to serve for a full blown
 *  AVI handling software (this would be much too complex)
 *  The only intention is to write out MJPEG encoded
 *  AVIs with sound and to be able to read them back again.
 *  These utilities should work with other types of codecs too, however.
 *
 *  Copyright (C) 1999 Rainer Johanni <Rainer@Johanni.de>

SDL Simple DirectMedia Layer http://www.libsdl.org
 *	Everybody at Loki Software, Inc. for their great contributions!
 
SHLIB Static Huffman encoder from Alexander Simakov <xander@entropyware.info>

JWebcamPlayer a java applet you can use to get the server stream trought an http
*	browser Copyrigth (C) 2005 Alvaro Salmador (nx5) (naplam33 at msn.com)

WHO CAN I BLAME?
================
Current maintainer and project lead: Michel Xhaard <mxhaard@free.fr>


DISCLAIMER:
-----------
Use this program as you see fit. We are not liable for damages or illegal use
of it whatsoever. You are on your own.
Use of this code is at your own risks. There is absolutely no waranty.
Have fun with it!

BASIC INSTALLATION:
-------------------
These are generic installation instructions.
You need sdl installed with the headers http://www.libsdl.org for spcaview
to compile all type:
make
to install in /usr/local/bin as root:
make install 
to clean the last compile work:
make clean
to compile and install as root only spcaserv (No dependance) 
make install_spcaserv
to compile and install only spcacat  (No dependance)
make install_spcacat

the package only need libsdl the jpeg encoder and decoder is now part of the package and allow to work with small footprint
for embedded device. libsdl provide the necessarry layer to use the package without a running Xserver and use the Linux framebuffer
instead :)
 
USAGE:
------
spcaview [-h -d -v -j -l -s -m -g -a -f -c -n -b -t -p -N -z] [-w IP:Port] [-i inputfile | -o outputfile]
spcaserv [-h -d -g -f -s] [-w Port]
spcacat [-h -d -g -f -s -p -N -P -o]
	Player mode 
	    -i inputfile   causes the program to read the avi image data from the specified file, rather than the video device. 
	    -w 	Address:Port read from Address xxx.xxx.xxx.xxx:Port
	    -w	Port server mode: write to the port
	    -m	widthxheight      use specified ouput size 
	Grabber Recorder mode
	    -h	show this usage message.
	    -d  device ask the driver to use specified video output device (/dev/video1)
	    
	    -o outputfile  causes the program to output avi  with image data received from the video device to the specified file.
	    -a	0 nosound
	    	1 input from default mixer input
	    -v grab in raw mode no output on the video screen fourcc RAWD
	    -j grab without display and decoding Ctrl C stop the stream
	    -z grab and compress videoyuv420p using DPSH encoder
	    -g grab with READ method instead of default MMAP
	    -f video format 	nothing YUV420P		fourcc I420
	    			jpg JPEG        	fourcc MJPG
	    			yuv YUV420P 		fourcc I420
				r16 RGB565 16bits	fourcc RGB2
				r24 RGB 24bits		fourcc RGB3
				r32 RGB 32bits		fourcc RGB4
				-v RAW data		fourcc RAWD
	    
	    -l	size [640x480] default is [320x240]
	    -s	widthxheight ask the driver for input size
	    -m	widthxheight      use specified ouput size 
	    -c  set the channel 0..3 video composite 6..9 Svideo
	    -n  select norme 0->pal, 1-> ntsc, 2-> secam
	    -b  select software autobrightness feature
	    -p X take a picture every X seconds
	    -p X && -o getPicture every X seconds and record in outputfile avi (spcaview)
	    -p X && -o getPicture every X seconds and record in SpcaPict.jpg (spcacat)
	    -N take N pictures and stop 
	    -P parport device (spcacat, spcaserv)
	    -t  display statistics 
	    
In Grabber mode the following keys are active
	key n	  brightness up
	key b     brightness down
	key v     color up
	key c     color down
	key x     contrast up
	key w	  contrast down
	key s	  screenshoot
	key j     autoexposure ON/OFF (with some webcams)
	key l	  Flicker filter 50/60/No
	key h     frame rate down
	key g     frame rate up
	key d     quality up (with some webcams)
	key f     quality down (with some webcams) 
In Grabber and Server mode without SDL Display
	CTRL C stop the stream and close the recorded file 

In Player Client mode the following keys are active
	key up    restart the video stream
	key down  stop the video stream
	key right display next frame
	key left  display last frame
	key space toggle Fullscreen and normal video mode
	key s     take a screen shoot in jpg or pnm.


Changelog
/************* Changelog spcaview *****************/
06/11/2006:
version 1.1.8
	Add bridge pac7311 pac207 vc0321 vc0323
06/11/2006:
version 1.1.7
	FIX broken s key feature !! Who know it was broken ??
	FIX 160x120 jpeg decoder for 422 stream
28/08/2006:
version 1.1.6
	Add Flicker filters 50Hz 60Hz No flickers Thanks Serge A. Suchkov
10/12/2005:
version 1.1.5rc2
	Add ioctl to spcav4l.c spcav4l.h
	implement callback to spcaserv and spcaview client
	change key to
	d quality up
	f quality down
	g framerate up
	h framerate down
	j autoexpo on/off
10/12/2005:
version 1.1.5rc1
	Add ioctl private management to spcaview (spcaserv and spcacat is not
	set ATM )
	quality
	autoexposure ON/off
	frame time interval
	Update Java Applet with a FIX from Andrea Partinico
	Update java Applet feature contrast and brightness can be set with the
	mouse now, Thanks to Andrea Partinico :)
	 
12/11/2005
version 1.1.4
	FIX -N parameters in spcacat 
	Update the java Applet
version 1.1.3
	spca5xx now can use a two framebuffers update spcaview 
	FIX problem with some camera the first frame is empty 
07/10/2005:
version 1.1.2
	Add a command key c on the client side server should send a 100ms pulse
	on pin 14
01/10/2005:
version 1.1.1
	Fix problem when parport is claim by another program disable use parport
24/09/2005:
version 1.1.0rc2
	Add "spcacat" tools take a picture in jpeg every x ms 
		store in a date/hours/Picture forms or overwrite the picture
		name
	look at spcacat -h for more info
	change makefile for spcacat
	make install_spcacat build and install the binary no depencies
23/09/2005:
	version 1.1.0rc1
	change makefile spcaserv can be compiled without libsdl with:
	make install_spcaserv
	Break server client compatibilty with previous version
	use packed struct to solve x86/x86_64 problem
	change the java Applet according the struct packed
	
18/09/2005:
	version 1.0.9
	FIX a bug in the bridge detection
	Add a lot of missing new bridge
14/06/2005:
	version 1.0.8
	Allow full screen in player mode 
	Fix only q should quit in grabber mode
	Add colors and exposure set get for the server messages
02/06/2005:
	version 1.0.7
	FIX some parport bugs with x and y messages
	FIX PPRELEASE on close
01/06/2005:
	version 1.0.6
	use parallel port and a small hardware available soon with spcaserv
	break the java applet sorry use the last revision or patch the applet:(
	client can now goes sleep and should be wakeup on an event on the server
	client can change brightness and contrast
	client can drive xy tracking camera set on the server available soon :)
05/05/2005:
	version 1.0.5:
	bugfix spca561 probe size
	Improve client server protocol
24/04/2005:
	version 1.0.4
	Add bridge Sonix sn9c102p jpeg without header
19/04/2005:
	version 1.0.3
	Client now display date and hours from the server
	Add a oneshoot feature with -N parameters -N can be combined with -p
	for example:
	spcaview -f jpg -s 320x240 -p 10 -N 2
	start spcaview take a picture 10 second after start then another one at
	20 seconds and exit 
	if -p is not set default delay is set to 1 second
	if -N is set to 0 :) then take one picture and exit 
27/03/2005:
	FIX bug segfault when the bridge are unknow
	version 1.0.2
26/03/2005:
	FIX -b segfault thanks Jos van Wolput for the feedback and test
	a lot of cleanup thanks Mark Studebaker	for the feedback and test
25/03/2005:
	Add the applet from Alvaro (nx5) :)
	FIX palette rgb565 and rgb32 in the server
	change version to 1.0.0
18/03/2005:
	FIX if client or server is down wait and shutdown the thread or the
	server.		
15/03/2005:
	big rewrite start :)
	Add resizer feature -m parameters can change aspect ratio :)
	Picture always goes in jpeg form
	record Picture as jpeg file or in an avi file 
	Picture now have is own thread
	split code player grabber server client
	FIX player in frame by frame mode wait instead poll a lot of CPU cycle
	saved :)
	For embedded device spcaview now have is own jpeg decoder and jpeg
	encoder (no more need jpeglib save 100ko memory space)
	FIX from Tyler Montbriand SDL_audiolib did not block if the sound device
	are opened in full duplex mode :)
01/01/2005
	Change UNIX itimer with signal used to SDL_timer 
	picture get with -p is now set by date/times 	
31/12/2004
	Version is set to 0.4 major change
	Add -j grab without SDL windows
	Add SDL_audioin instead sndlib
		SDL_audioin is compiled inside spcaview so no need of external dependancies
		Rewrite the sound support works with low frame rate camera 5 fp/s and hight rate 30 fp/s	
	Add -p options get picture every N seconds Thanks David Lecombier and Stas Bevc
	Add -t Print statistic on the console
21/12/2004
	Add Conexant cx11646 bridge
01/10/2004
	Add alternate resolution for the zc30x chips (available only for rgb yuv
	palette)
	Change version number spcaview-01102004 become spcaview-20041001
26/09/2004
	Add support for the zc30x chips (jpeg 4.2.2)		
01/09/2004
	Add support for the spca536, Our Sonix and Zr364xx Camera driver
	Make -f jpg work with true jpeg stream get from the kernel
23/08/2004
	Fix some problem with autoadjust the initial brightness is get from the cam now
	Adjust MAXBRIGHT to 50000 to avoid saturation with some cam
	restore the statitistic feature when autobrightness is not set
	and autobrightness value when set
17/08/2004
	Add autobrightness feature from Gihan Marasingha
	change version 0.29
20/05/2004
	Add probe function spca5xx camera report palette and size available
	bridge and Camera name.	
15/04/2004
	Add -n norme parameters 0 ->PAL || 1 -> NTSC || 2 -> Secam
	Add -c Channel [0..3] VideoComposite [6..9] S-Video
	Change version 0.27
13/03/2004
	Clean up Statistic
	Change version 0.26 
26/02/2004
	Add compressor for yuv420p stream "Differential Pixel Static Huffman" encoder decoder
	option -z that allow compression without loss on yuv stream (~50% in size)
21/01/2004
	Add screenshot when the cam stream only press the "s" key when you have focus 
	on the windows grabber.The Picture goes in your current directory.

jconfig.h
#ifndef JCONFIG_H
#define JCONFIG_H

#define IMAGE_WIDTH 320 
#define IMAGE_HEIGHT 240

#define VIDEO_PALETTE_RAW_JPEG  20
#define VIDEO_PALETTE_JPEG  21
#define SWAP_RGB 1
int bpp;
/* our own ioctl */
struct video_param {
	int chg_para;
#define CHGABRIGHT 1
#define CHGQUALITY 2
#define CHGLIGHTFREQ 3
#define CHGTINTER  4
	__u8 autobright;
	__u8 quality;
	__u16 time_interval;
	__u8 light_freq;
};

/* Our private ioctl */
#define SPCAGVIDIOPARAM _IOR('v',BASE_VIDIOCPRIVATE + 1,struct video_param)
#define SPCASVIDIOPARAM _IOW('v',BASE_VIDIOCPRIVATE + 2,struct video_param)

/* Camera type jpeg yuvy yyuv yuyv grey gbrg*/
enum {  
	JPEG = 0,
	YUVY,
	YYUV,
	YUYV,
	GREY,
	GBRG,
	UNOW,
};
#if 0
#define BRIDGE_SPCA505 0
#define BRIDGE_SPCA506 1
#define BRIDGE_SPCA501 2
#define BRIDGE_SPCA508 3
#define BRIDGE_SPCA504 4
#define BRIDGE_SPCA500 5
#define BRIDGE_SPCA504B 6
#define BRIDGE_SPCA533 7
#define BRIDGE_SPCA504C 8
#define BRIDGE_SPCA561 9
#define BRIDGE_SPCA536 10
#define BRIDGE_SONIX 11
#define BRIDGE_ZR364XX 12
#define BRIDGE_ZC3XX 13
#define BRIDGE_CX11646 14
#define BRIDGE_TV8532 15
#define BRIDGE_ETOMS 16
#define BRIDGE_SN9CXXX 17
#define BRIDGE_MR97311 18
#define MAX_BRIDGE 19

#endif

enum {
	BRIDGE_SPCA505 = 0,
        BRIDGE_SPCA506,
	BRIDGE_SPCA501,
	BRIDGE_SPCA508,
	BRIDGE_SPCA504,
	BRIDGE_SPCA500,
	BRIDGE_SPCA504B,
	BRIDGE_SPCA533,
	BRIDGE_SPCA504C,
	BRIDGE_SPCA561,
	BRIDGE_SPCA536,
	BRIDGE_SONIX,
	BRIDGE_ZC3XX,
	BRIDGE_CX11646,
	BRIDGE_TV8532,
	BRIDGE_ETOMS,
	BRIDGE_SN9CXXX,
	BRIDGE_MR97311,
	BRIDGE_PAC207,
	BRIDGE_VC0321,
	BRIDGE_VC0323,
	BRIDGE_PAC7311,
	BRIDGE_UNKNOW,
	MAX_BRIDGE,
};
struct palette_list {
	int num;
	const char *name;
};

struct bridge_list {
	int num;
	const char *name;
};
#if 0
static struct bridge_list Blist[]={

	{BRIDGE_SPCA505,"SPCA505"},
	{BRIDGE_SPCA506,"SPCA506"},
	{BRIDGE_SPCA501,"SPCA501"},
	{BRIDGE_SPCA508,"SPCA508"},
	{BRIDGE_SPCA504,"SPCA504"},
	{BRIDGE_SPCA500,"SPCA500"},
	{BRIDGE_SPCA504B,"SPCA504B"},
	{BRIDGE_SPCA533,"SPCA533"},
	{BRIDGE_SPCA504C,"SPCA504C"},
	{BRIDGE_SPCA561,"SPCA561"},
	{BRIDGE_SPCA536,"SPCA536"},
	{BRIDGE_SONIX,"SN9C102"},
	{BRIDGE_ZR364XX,"ZR364XX"},
	{BRIDGE_ZC3XX,"ZC301-2"},
	{BRIDGE_CX11646,"CX11646"},
	{BRIDGE_TV8532, "TV8532"},
  	{BRIDGE_ETOMS,"ET61XX51"},
	{BRIDGE_SN9CXXX,"SN9CXXX"},
	{BRIDGE_MR97311,"MR97311"},
	{-1,NULL}
};
#endif

static struct bridge_list Blist[]={

	{BRIDGE_SPCA505,"SPCA505"},
	{BRIDGE_SPCA506,"SPCA506"},
	{BRIDGE_SPCA501,"SPCA501"},
	{BRIDGE_SPCA508,"SPCA508"},
	{BRIDGE_SPCA504,"SPCA504"},
	{BRIDGE_SPCA500,"SPCA500"},
	{BRIDGE_SPCA504B,"SPCA504B"},
	{BRIDGE_SPCA533,"SPCA533"},
	{BRIDGE_SPCA504C,"SPCA504C"},
	{BRIDGE_SPCA561,"SPCA561"},
	{BRIDGE_SPCA536,"SPCA536"},
	{BRIDGE_SONIX,"SN9C102"},
	{BRIDGE_ZC3XX,"ZC301-2"},
	{BRIDGE_CX11646,"CX11646"},
	{BRIDGE_TV8532,"TV8532"},
	{BRIDGE_ETOMS,"ET61XX51"},
	{BRIDGE_SN9CXXX,"SN9CXXX"},
	{BRIDGE_MR97311,"MR97311"},
	{BRIDGE_PAC207, "PAC207BCA"},
	{BRIDGE_VC0321, "VC0321"},
	{BRIDGE_VC0323, "VC0323"},
	{BRIDGE_PAC7311, "PAC7311"},
	{BRIDGE_UNKNOW,"UNKNOW"},
	{-1,NULL}
};
/* Camera type jpeg yuvy yyuv yuyv grey gbrg*/
static struct palette_list Plist[] ={
	{JPEG,"JPEG"},
	{YUVY,"YUVY"},
	{YYUV,"YYUV"},
	{YUYV,"YUYV"},
	{GREY,"GREY"},
	{GBRG,"GBRG"},
	{-1,NULL}
};

#define GET_EXT_MODES(bridge) (\
	(bridge) == BRIDGE_SPCA500 ? spca500_ext_modes : \
	(bridge) == BRIDGE_SPCA501 ? spca501_ext_modes : \
	(bridge) == BRIDGE_SPCA504 ? spca504_ext_modes : \
	(bridge) == BRIDGE_SPCA504B ? spca504_ext_modes : \
	(bridge) == BRIDGE_SPCA504C ? spca504_pccam600_ext_modes : \
	(bridge) == BRIDGE_SPCA506 ? spca506_ext_modes : \
	(bridge) == BRIDGE_SPCA508 ? spca508_ext_modes : \
	(bridge) == BRIDGE_SPCA533 ? spca533_ext_modes : \
	(bridge) == BRIDGE_SPCA561 ? spca561_ext_modes : \
	(bridge) == BRIDGE_SPCA536 ? spca536_ext_modes : \
	(bridge) == BRIDGE_SONIX ? sonix_ext_modes : \
	(bridge) == BRIDGE_ZC3XX ? zc3xx_ext_modes : \
	(bridge) == BRIDGE_CX11646 ? cx11646_ext_modes : \
	(bridge) == BRIDGE_SN9CXXX ? sn9c102p_ext_modes : \
	spca50x_ext_modes)

static unsigned int sn9c102p_ext_modes[][3] = {
	/* x , y , Code */
	{640, 480, 0x00},
	{352, 288, 0x10},	
	{320, 240, 0x01},
	{176, 144, 0x11},
	{0, 0, 0}
};
static unsigned int cx11646_ext_modes[][3] = {
	/* x , y , Code */
	{640, 480, 0x00},
	{352, 288, 0x01},	
	{320, 240, 0x02},
	{176, 144, 0x03},
	{0, 0, 0}
};
static unsigned int zc3xx_ext_modes[][3] = {
	/* x , y , Code */
	{640, 480, 0x00},
	{352, 288, 0x10},	
	{320, 240, 0x01},
	{176, 144, 0x11},
	{0, 0, 0}
};

static unsigned int zr364xx_ext_modes[][3] = {
	/* x , y , Code */
	{640, 480, 0x00},	
	{320, 240, 0x01},
	{160, 120, 0x02},
	{0, 0, 0}
};

static unsigned int sonix_ext_modes[][3] = {
	/* x , y , Code */
	{640, 480, 0x00},	
	{320, 240, 0x01},
	{160, 120, 0x02},
	{0, 0, 0}
};

static unsigned int spca500_ext_modes[][3] = {
	/* x , y , Standard*/
	{640, 480, 0x00},
	{352, 288, 0x00},
	{320, 240, 0x00},
	{176, 144, 0x00},
	{0, 0, 0}
};

static unsigned int spca501_ext_modes[][3] = {
	
	{640, 480, 0x00},
	{352, 288, 0x00},
	{320, 240, 0x00},
	{176, 144, 0x00},
	{160, 120, 0x00},
	{0, 0, 0}
};

static unsigned int spca504_ext_modes[][3] =
{	
	{ 640, 480, 0x01},
	{ 384, 288, 0x11},
	{ 352, 288, 0x11},
	{ 320, 240, 0x02},
	{ 192, 144, 0x12},
	{ 176, 144, 0x12},
	{ 0, 0, 0}
};

 
static unsigned int spca504_pccam600_ext_modes[][3] =
{	
	{ 1024, 768, 0x00}, 
	{ 640,  480, 0x01},
	{ 352,  288, 0x02},
	{ 320,  240, 0x03},
	{ 176,  144, 0x04},
	{ 0, 0, 0}
};


static unsigned int spca506_ext_modes[][3] =
{	
	{ 640, 480, 0x00},
	{ 352, 288, 0x01},
	{ 320, 240, 0x02},
	{ 176, 144, 0x04}, 
	{ 160, 120, 0x05},
	{ 0, 0, 0}
};

static unsigned int spca508_ext_modes[][3] =
{	
	{ 352, 288, 0x00},
	{ 320, 240, 0x01},
	{ 176, 144, 0x02},
	{ 160, 120, 0x03},
	{ 0, 0, 0}
};

static unsigned int spca533_ext_modes[][3] =
{	
	{ 640, 480, 0x41},
	{ 464, 480, 0x01},//PocketDVII unscaled resolution aspect ratio need to expand x axis
	{ 464, 352, 0x01},//Gsmart LCD3 feature good aspect ratio
	{ 384, 288, 0x11},
	{ 352, 288, 0x11},
	{ 320, 240, 0x02},
	{ 192, 144, 0x12},
	{ 176, 144, 0x12},
	{ 0, 0, 0}
};

static unsigned int spca561_ext_modes[][3] =
{
	{ 352, 288, 0x00},
	{ 320, 240, 0x01},
	{ 176, 144, 0x02},
	{ 160, 120, 0x03},
	{ 0, 0, 0}
};

static unsigned int spca536_ext_modes[][3] =
{
	/* x , y , Code */
	{ 464, 480, 0x01},
	{ 464, 352, 0x01},
	{ 384, 288, 0x11},
	{ 352, 288, 0x11},
	{ 320, 240, 0x02},
	{ 192, 144, 0x12},
	{ 176, 144, 0x12},	
	{ 0, 0, 0}
};

static unsigned int spca50x_ext_modes[][3] =
{	
	{ 640, 480, 0x00},
	{ 352, 288, 0x01},
	{ 320, 240, 0x02}, 
	{ 176, 144, 0x04}, 
	{ 160, 120, 0x05}, 
	{ 0, 0, 0}
};


#endif /* JCONFIG_H */
